contracts/BonusRewards.sol

```diff
@@ -3,11 +3,361 @@
 pragma solidity ^0.8.0;
 
 import "./utils/Ownable.sol";
+import "./utils/ReentrancyGuard.sol";
+import "./utils/SafeERC20.sol";
 import "./interfaces/IBonusRewards.sol";
 
 /**
  * @title Cover Protocol Bonus Token Rewards contract
  * @author crypto-pumpkin
+ * @notice ETH is not allowed to be an bonus token, use wETH instead
+ * @notice We support multiple bonus tokens for each pool. However, each pool will have 1 bonus token normally, may have 2 in rare cases
  */
-contract BonusRewards is Ownable {
+contract BonusRewards is IBonusRewards, Ownable, ReentrancyGuard {
+  using SafeERC20 for IERC20;
+
+  bool public paused;
+  uint256 private constant WEEK = 7 days;
+  // help calculate rewards/bonus PerToken only. 1e12 will allow meaningful $1 deposit in a $1bn pool  
+  uint256 private constant CAL_MULTIPLIER = 1e12;
+  // use array to allow convinient replacement. Size of responders should be very small to 0 till a reputible responder multi-sig within DeFi or Yearn ecosystem is established
+  address[] private responders;
+  address[] private poolList;
+  // lpToken => Pool
+  mapping(address => Pool) private pools;
+  // lpToken => User address => User data
+  mapping(address => mapping(address => User)) private users;
+  // use array to allow convinient replacement. Size of Authorizers should be very small (one or two partner addresses for the pool and bonus)
+  // lpToken => bonus token => [] allowed authorizers to add bonus tokens
+  mapping(address => mapping(address => address[])) private allowedTokenAuthorizers;
+  // bonusTokenAddr => 1, used to avoid collecting bonus token when not ready
+  mapping(address => uint8) private bonusTokenAddrMap;
+
+  modifier notPaused() {
+    require(!paused, "BonusRewards: paused");
+    _;
+  }
+
+  function claimRewardsForPools(address[] calldata _lpTokens) external override nonReentrant notPaused {
+    for (uint256 i = 0; i < _lpTokens.length; i++) {
+      address lpToken = _lpTokens[i];
+      User memory user = users[lpToken][msg.sender];
+      if (user.amount == 0) continue;
+      _updatePool(lpToken);
+      _claimRewards(lpToken, user);
+      _updateUserWriteoffs(lpToken);
+    }
+  }
+
+  function deposit(address _lpToken, uint256 _amount) external override nonReentrant notPaused {
```

Comment:
This doesn't support deflationary tokens, is that intended?

Reply:
Good catch. Dont think we need. But we should handle it. Will update